File: src/main/java/network/reflected/rfnetapi/RfnetAPI.java
package network.reflected.rfnetapi;

import com.google.common.io.ByteArrayDataOutput;
import com.google.common.io.ByteStreams;
import com.grinderwolf.swm.api.loaders.SlimeLoader;
import com.grinderwolf.swm.api.world.SlimeWorld;
import com.grinderwolf.swm.api.world.properties.SlimeProperties;
import com.grinderwolf.swm.api.world.properties.SlimePropertyMap;
import com.grinderwolf.swm.plugin.SWMPlugin;
import lombok.Getter;
import org.apache.commons.io.IOUtils;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.plugin.java.JavaPlugin;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.net.URL;
import java.net.URLConnection;
import java.util.Base64;
import java.util.List;
import java.util.Random;
import java.util.logging.Level;

public final class RfnetAPI extends JavaPlugin implements Listener {
    private final int ver = 10; // The current version
    private boolean disabledForUpdate = false;

    @Getter private ReflectedAPI api;
    @Getter private final ServerConfig serverConfig = new ServerConfig();
    @Getter private final Database database = new Database(serverConfig);
    @Getter private String loadedMapName = "world";
    @Getter private boolean minigameWorld = false;

    @Override
    public void onEnable() {
        // Init the API and let any waiting plugins know that it's ready now.
        api = new ReflectedAPI(this);
        ReflectedAPI.checkCallbacks();

        // If something is wrong with the config, shutdown the server, since it won't be connectable.
        if (!serverConfig.isValid()) {
            getLogger().log(Level.SEVERE, serverConfig.whatsMissing());
            getServer().shutdown();
        }

        // If the database isn't connected, its either having issues or not available.
        // Since this server won't be connectable, shut it down.
        if (!database.isConnected()) {
            getLogger().log(Level.SEVERE, "Could not connect to database! Look in console for errors.");
            getServer().shutdown();
        }

        // Setup a plugin messaging channel
        getServer().getMessenger().registerOutgoingPluginChannel(this, "BungeeCord");

        // Register this class as an event listener
        getServer().getPluginManager().registerEvents(this, this);

        // Register other API's events
        ReflectedAPI.get(api -> {
            getServer().getPluginManager().registerEvents(api.getCommandProvider(), this);
            getServer().getPluginManager().registerEvents(api.getPurchaseAPI(), this);
        });

        // Setup default commands, available on every server
        DefaultCommands.initialize();

        // Some stuff should be run AFTER the server has fully loaded.
        getServer().getScheduler().runTaskLater(this, () -> {
            // Load the worlds as defined in the config
            // Start by setting up the SWM plugin
            getLogger().info("Getting SWM plugin.");
            try {
                if (Bukkit.getPluginManager().getPlugin("SlimeWorldManager") == null) {
                    throw new NoClassDefFoundError();
                }
                SWMPlugin slime = SWMPlugin.getInstance();
                if (slime != null) { // If slime plugin can't be found, we're not on a minigame server.
                    getLogger().info("Configuring worlds...");
                    SlimeLoader slimeMongoLoader = slime.getLoader("mongodb");

                    // Now set some properties of the world
                    SlimePropertyMap mapProperties = new SlimePropertyMap();
                    mapProperties.setValue(SlimeProperties.DIFFICULTY, "normal");
                    mapProperties.setValue(SlimeProperties.SPAWN_X, 0);
                    mapProperties.setValue(SlimeProperties.SPAWN_Y, 64);
                    mapProperties.setValue(SlimeProperties.SPAWN_Z, 0);
                    mapProperties.setValue(SlimeProperties.WORLD_TYPE, "FLAT"); // removes void effect at lower y levels

                    // And finally, find out which world to load.
                    getLogger().info("Looking for worlds to load.");
                    try {
                        SlimeWorld loadedMap;
                        if (serverConfig.getMaps().size() == 1) { // There is only one map to choose
                            // Also that magic boolean before mapProperties is whether it's read only. No slime worlds we load
                            // would be good to make editable (they are all minigame maps)
                            loadedMap = slime.loadWorld(slimeMongoLoader, serverConfig.getMaps().get(0), true, mapProperties);
                        } else { // A map must be chosen at random.
                            List<String> maps = serverConfig.getMaps();
                            Random rand = new Random();
                            // See above for what the magic boolean is
                            loadedMap = slime.loadWorld(slimeMongoLoader, maps.get(rand.nextInt(maps.size())), true, mapProperties);
                        }

                        getLogger().info("Loading " + loadedMap.getName());
                        slime.generateWorld(loadedMap);
                        loadedMapName = loadedMap.getName();
                        minigameWorld = true;
                    } catch (Exception e) {
                        // If the map fails to load, there's nothing to connect to, so stop the server.
                        e.printStackTrace();
                        getLogger().log(Level.SEVERE, "Error loading a map! The server has to shut down!");
                        getServer().shutdown();
                    }
                }
            } catch (NoClassDefFoundError e) {
                getLogger().info("An error occured when attempting to load SWM, so minigame world support has been disabled.");
            }

            // Add this server's information to Redis for ServerDiscovery.
            database.setAvailable(true);
        }, 1); // 1 tick, so waits until the server is fully started (started ticking)
    }

    @Override
    public void onDisable() {
        if (!disabledForUpdate) {
            updateCheck();
            genericDisable();
        }
    }

    public void genericDisable() {
        // Remove this server from the list of ones that are connectable
        database.setAvailable(false);
        // And then close the connections to the database
        // so we don't overload them.
        database.close();
    }

    // Sends a plugin message to ServerDiscovery running on bungee.
    public void sendPlayer(Player player, String archetype) {
        if (!api.getPurchaseAPI().isAuthenticating(player) || archetype.equals("lobby")) {  // If the player is busy making a purchase or changing a password, don't interrupt it.
            ByteArrayDataOutput out = ByteStreams.newDataOutput();

            // See the spec for this in ServerDiscovery
            out.writeUTF("send:" + player.getUniqueId() + ":" + archetype);

            player.sendPluginMessage(this, "BungeeCord", out.toByteArray());
        }
    }

    @EventHandler(priority = EventPriority.LOWEST)
    private void playerJoin(PlayerJoinEvent event) {
        if (isMinigameWorld()) {
            Location location = event.getPlayer().getLocation();
            location.setWorld(Bukkit.getWorld(loadedMapName));
            event.getPlayer().teleport(location);
        }
        database.updatePlayerCount(Bukkit.getOnlinePlayers().size());
    }

    @EventHandler
    private void playerQuit(PlayerQuitEvent event) {
        database.updatePlayerCount(Bukkit.getOnlinePlayers().size() - 1);
    }


    public void restart() {
        disabledForUpdate = true;

        genericDisable();

        // Send everybody to another server
        for (Player player : Bukkit.getOnlinePlayers()) {
            ReflectedAPI.get().sendPlayer(player, ReflectedAPI.get().getPlugin().getServerConfig().getArchetype());
        }

        updateCheck();

        // Wait one second so players don't get Server Closed before being sent back to lobby
        Bukkit.getScheduler().runTaskLater(this, () -> {
            Runtime runtime = Runtime.getRuntime();
            runtime.addShutdownHook(new Thread(() -> {
                ProcessBuilder processBuilder = new ProcessBuilder("nohup", "sh", "restart.sh");
                try {
                    processBuilder.directory(new File("."));
                    processBuilder.redirectErrorStream(false);
                    processBuilder.start();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }));
            Bukkit.shutdown();
        }, 20);
    }

    public void updateCheck() {
        System.out.println("Checking for updates...");
        // Check for updates
        int nextVer = ver + 1;
        String updateOneLiner = "/bin/wget  --http-user=$(/bin/cat ~/.spaceauth/user) --http-password=$(/bin/cat ~/.spaceauth/passwd)";
        ProcessBuilder updProcessBuilder = new ProcessBuilder("/bin/bash", updateOneLiner);
        // Credentials for a read-only user
        String username = "31faf87b-0584-449b-b5b4-542b711fedfd";
        String password = "eyJhbGciOiJSUzUxMiJ9.eyJzdWIiOiIzMWZhZjg3Yi0wNTg0LTQ0OWItYjViNC01NDJiNzExZmVkZmQiLCJhdWQiOiIzMWZhZjg3Yi0wNTg0LTQ0OWItYjViNC01NDJiNzExZmVkZmQiLCJvcmdEb21haW4iOiJyZWZsZWN0ZWRuZXR3b3JrIiwibmFtZSI6InNlcnZlci11cGRhdGVyIiwiaXNzIjoiaHR0cHM6XC9cL2pldGJyYWlucy5zcGFjZSIsInBlcm1fdG9rZW4iOiIzZE9ueW4zVjY4U0oiLCJwcmluY2lwYWxfdHlwZSI6IlNFUlZJQ0UiLCJpYXQiOjE2MjQ0Nzk3MDB9.BmoDM8WFtadwvF9506wpDugq7yNtaiCyWIzfX-cC4dGYSQgtTJrISH85fYI0ukD8E4_xFN74xmwa6Pc1gI6kCYICFkk1YIREsdAS08m4KCtAM4iK5YFAD2aodlIOgGVVon7EEqTM8KHBCjIyEVk9R-Vj8tLi37j4cnubfo7VkMk";

        File download = new File("./plugins/RfnetAPI-" + nextVer + ".jar");
        try {
            FileOutputStream downloadStream = new FileOutputStream(download);
            String basicAuthenticationEncoded = Base64.getEncoder().encodeToString((username + ":" + password).getBytes("UTF-8"));
            URL url = new URL("https://maven.pkg.jetbrains.space/reflectednetwork/p/internalapi/maven/network/reflected/RfnetAPI/" + nextVer + "/RfnetAPI-" + nextVer + ".jar");
            URLConnection urlConnection = url.openConnection();
            urlConnection.setRequestProperty("Authorization", "Basic " + basicAuthenticationEncoded);
            IOUtils.copy(urlConnection.getInputStream(), downloadStream);

            new File("./plugins/RfnetAPI-" + ver + ".jar").deleteOnExit();
            System.out.println("Update complete!");
        } catch (FileNotFoundException e) {
            System.out.println("No update found!");
            download.delete();
        } catch (Exception e) {
            e.printStackTrace();
            System.out.println("Update failed! See error above!");
        }
    }
}

File: src/main/java/com/reflectednetwork/rfnetapi/RfnetAPI.kt
package com.reflectednetwork.rfnetapi
import com.google.common.io.ByteStreams
import com.google.gson.JsonObject
import com.reflectednetwork.rfnetapi.async.async
import com.reflectednetwork.rfnetapi.bugs.ExceptionDispensary
import com.reflectednetwork.rfnetapi.medallions.MedallionAPI
import com.reflectednetwork.rfnetapi.modtools.ModCommands
import com.reflectednetwork.rfnetapi.modtools.ModEvents
import com.reflectednetwork.rfnetapi.permissions.PermissionAPI
import com.reflectednetwork.rfnetapi.permissions.PermissionCommands
import com.reflectednetwork.rfnetapi.purchases.PurchaseEvents
import com.reflectednetwork.rfnetapi.purchases.PurchaseGUI
import io.ktor.client.*
import io.ktor.client.features.json.*
import io.ktor.client.request.*
import kotlinx.coroutines.runBlocking
import org.apache.commons.io.IOUtils
import org.bstats.bukkit.Metrics
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.event.Listener
import org.bukkit.plugin.java.JavaPlugin
import java.io.File
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException
import java.net.URL
import java.net.URLConnection
import java.nio.file.Files
import java.nio.file.Paths
import java.util.*
import java.util.logging.Level

class RfnetAPI : JavaPlugin(), Listener {
    val ver = description.version.toInt()
    private var disabledForUpdate = false

    private val nextVer = ver + 1
    // Credentials for a read-only user
    private val spaceUser = "31faf87b-0584-449b-b5b4-542b711fedfd"
    private val spacePassword = "eyJhbGciOiJSUzUxMiJ9.eyJzdWIiOiIzMWZhZjg3Yi0wNTg0LTQ0OWItYjViNC01NDJiNzExZmVkZmQiLCJ" +
            "hdWQiOiIzMWZhZjg3Yi0wNTg0LTQ0OWItYjViNC01NDJiNzExZmVkZmQiLCJvcmdEb21haW4iOiJyZWZsZWN0ZWRuZXR3b3JrIiwibmF" +
            "tZSI6InNlcnZlci11cGRhdGVyIiwiaXNzIjoiaHR0cHM6XC9cL2pldGJyYWlucy5zcGFjZSIsInBlcm1fdG9rZW4iOiIzZE9ueW4zVjY" +
            "4U0oiLCJwcmluY2lwYWxfdHlwZSI6IlNFUlZJQ0UiLCJpYXQiOjE2MjQ0Nzk3MDB9.BmoDM8WFtadwvF9506wpDugq7yNtaiCyWIzfX-" +
            "cC4dGYSQgtTJrISH85fYI0ukD8E4_xFN74xmwa6Pc1gI6kCYICFkk1YIREsdAS08m4KCtAM4iK5YFAD2aodlIOgGVVon7EEqTM8KHBCj" +
            "IyEVk9R-Vj8tLi37j4cnubfo7VkMk"

    var api: ReflectedAPI? = null
    val serverConfig = ServerConfig()
    val database = Database(serverConfig)
    var minigameWorld = false
    private lateinit var permissionAPI: PermissionAPI
    var ghostMode = false

    override fun onEnable() {
        try {
            // Init the API and let any waiting plugins know that it's ready now.
            api = ReflectedAPI(this)

            if (updateCheck()) {
                async {
                    logger.log(Level.SEVERE, "Plugin or dependencies out of date! Will restart server shortly.")
                }
                Bukkit.getScheduler().runTaskLater(this, Runnable { restart() }, 20)
                return
            }

            if (Bukkit.getWorldContainer().list()?.contains("RFNET_GHOST_MODE") == true) {
                ghostMode = true
                GhostModeManager.enable(this)
                return
            }

            // Because of really dumb dependency issues, we've gotta do this
            WorldPluginInterface.plugin = this



            permissionAPI = PermissionAPI(this)
            PermissionCommands.setupCommands()

            ModCommands.initCommands()

            // If something is wrong with the config, shutdown the server, since it won't be connectable.
            if (!serverConfig.isValid()) {
                logger.log(Level.SEVERE, serverConfig.whatsMissing())
                server.shutdown()
            }

            // Setup a plugin messaging channel
            server.messenger.registerOutgoingPluginChannel(this, "BungeeCord")

            // Register other API's events
            server.pluginManager.registerEvents(ReflectedAPI.get().commandProvider, this)
            server.pluginManager.registerEvents(ReflectedAPI.get().purchaseAPI, this)

            server.pluginManager.registerEvents(PurchaseGUI, this)
            server.pluginManager.registerEvents(PurchaseEvents, this)
            server.pluginManager.registerEvents(ExceptionDispensary, this)
            server.pluginManager.registerEvents(MedallionAPI, this)
            server.pluginManager.registerEvents(JoinEventWorkaround, this)
            server.pluginManager.registerEvents(PlayerCountEvents, this)
            server.pluginManager.registerEvents(ModEvents, this)
            server.pluginManager.registerEvents(permissionAPI, this)

            // Check online receipts on occasion.
            server.scheduler.runTaskTimerAsynchronously(this, Runnable {
                try {
                    for (player in server.onlinePlayers) {
                        PurchaseEvents.checkReceipts(player)
                    }
                } catch (e: Exception) {
                    ExceptionDispensary.report(e, "checking receipts")
                }
            }, 400, 400)

            // Also, check for updates if no one is online
            server.scheduler.runTaskTimerAsynchronously(this, Runnable {
                try {
                    if (Bukkit.getOnlinePlayers().isEmpty() && updateCheck()) {
                        restart()
                    }
                } catch (e: Exception) {
                    ExceptionDispensary.report(e, "checking for updates")
                }
            }, 400, 400)

            // Setup default commands, available on every server
            DefaultCommands.initialize()

            // Some stuff should be run AFTER the server has fully loaded.
            server.scheduler.runTaskLater(this, Runnable {
                // Add this server's information to Redis for ServerDiscovery.
                database.setAvailable(true)
            }, 1) // 1 tick, so waits until the server is fully started (started ticking)

            // Stats
            Metrics(this, 12072)
        } catch (e: Exception) {
            ExceptionDispensary.report(e, "enabling plugin")
        }
    }

    override fun onDisable() {
        try {
            if (ghostMode) {
                GhostModeManager.disable()
                return
            }

            if (!disabledForUpdate) {

                // Remove this server from the list of ones that are connectable
                database.setAvailable(false)
                database.updatePlayerCount(0)
                // Close the connections to the database
                // so we don't overload them.
                database.close()
                // and then update the server
                runBlocking { update() }
            }
        } catch (e: Exception) {
            ExceptionDispensary.report(e, "disabling plugin")
        }
    }

    // Sends a plugin message to ServerDiscovery running on bungee.
    fun sendPlayer(player: Player, archetype: String) {
        try {
            @Suppress("UnstableApiUsage") val out = ByteStreams.newDataOutput()

            // See the spec for this in ServerDiscovery
            out.writeUTF("send:" + player.uniqueId + ":" + archetype)
            player.sendPluginMessage(this, "BungeeCord", out.toByteArray())
        } catch (e: Exception) {
            ExceptionDispensary.report(e, "sending player")
        }
    }

    fun restart() {
        try {
            disabledForUpdate = true

            // Remove this server from the list of ones that are connectable
            database.setAvailable(false)
            database.updatePlayerCount(0)

            // Send everybody to another server
            for (player in Bukkit.getOnlinePlayers()) {
                sendPlayer(player, serverConfig.archetype)
            }

            // Close the connections to the database
            // so we don't overload them.
            database.close()

            // And then update the server
            runBlocking { update() }

            // Wait one second so players don't get Server Closed before being sent back to lobby
            Bukkit.getScheduler().runTaskLater(this, Runnable {
                try {
                    Runtime.getRuntime().addShutdownHook(Thread {
                        val processBuilder = ProcessBuilder("nohup", "sh", "restart.sh")
                        try {
                            processBuilder.directory(File("."))
                            processBuilder.redirectErrorStream(false)
                            processBuilder.start()
                        } catch (e: IOException) {
                            ExceptionDispensary.report(e, "restarting")
                        }
                    })
                    Bukkit.shutdown()
                } catch (e: Exception) {
                    ExceptionDispensary.report(e, "shutting down for restart")
                }
            }, 20)
        } catch (e: Exception) {
            ExceptionDispensary.report(e, "restarting")
        }
    }

    fun updateCheck(): Boolean {
        val worldLoaderJar = File("./plugins/RFNETAPI_WorldLoader-1.jar")
        val protocolLibJar = File("./plugins/ProtocolLib-4.7.0.jar")
        val cclibCompatJar = File("./plugins/CCLib-1.0-SNAPSHOT.jar")

        if (!worldLoaderJar.exists() || !protocolLibJar.exists() || !cclibCompatJar.exists() || (!server.allowFlight || server.onlineMode)) {
            return true
        }

        return try {
            getSpaceConnection().getInputStream().readAllBytes()
            true
        }  catch (e: FileNotFoundException) {
            false
        } catch (e: IOException) { // We got a 404 meaning the file doesn't exist
            if (e.message?.contains("404") != true) throw e
            false
        }
    }

    private suspend fun update() {
        try {
            println("UPDATING > API")
            // Check for updates
            val pluginUpdateFile = File("./plugins/RfnetAPI-$nextVer.jar")
            try {
                val downloadStream = FileOutputStream(pluginUpdateFile)
                IOUtils.copy(getSpaceConnection().getInputStream(), downloadStream)
                File("./plugins/RfnetAPI-$ver.jar").deleteOnExit()
                println("--> Plugin updated")
            }  catch (e: FileNotFoundException) {
                pluginUpdateFile.delete()
            } catch (e: IOException) { // We got a 404 meaning the file doesn't exist
                if (e.message?.contains("404") != true) throw e
                pluginUpdateFile.delete()
            }

            println("UPDATING > Configuration")
            if (!server.allowFlight || server.onlineMode) {
                val properties = Files.lines(Paths.get("server.properties"))
                val newProperties = File("server.properties~").outputStream()

                properties.map {
                    when (it) {
                        "allow-flight=false" -> "allow-flight=true"
                        "network-compression-threshold=256" -> "network-compression-threshold=-1"
                        "view-distance=10" -> "view-distance=5"
                        "allow-nether=true" -> "allow-nether=${serverConfig.archetype == "survival"}"
                        "online-mode=true" -> "online-mode=false"
                        "server-port=25565" -> "server-port=${serverConfig.address.split(":").last()}"
                        else -> it
                    }
                }.forEach { line ->
                    newProperties.write("$line\n".encodeToByteArray())
                }

                Runtime.getRuntime().addShutdownHook(Thread {
                    val propertiesFile = File("server.properties")
                    propertiesFile.delete()
                    Files.move(Paths.get("server.properties~"), Paths.get("server.properties"))
                })

                println("--> Configuration updates applied")
            }

            println("UPDATING > Dependencies")
            download("https://github.com/dmulloy2/ProtocolLib/releases/download/4.7.0/ProtocolLib.jar", "ProtocolLib-4.7.0")
            download("https://www.dropbox.com/s/od0syes7xubidh3/RFNETAPI_WorldLoader-1.jar?dl=1", "RFNETAPI_WorldLoader-1")
            download("https://www.dropbox.com/s/v569132ztwnfqbr/CCLib-1.0-SNAPSHOT.jar?dl=1", "CCLib-1.0-SNAPSHOT")

            println("UPDATING > Core")
            val paperjar = File("./paper_1.17.1.jar")
            val client = HttpClient() {
                install(JsonFeature) {
                    serializer = GsonSerializer()
                }
            }

            val versionResponse: JsonObject = client.request("https://papermc.io/api/v2/projects/paper/versions/1.17.1")
            val latest = versionResponse.getAsJsonArray("builds").maxOf { it.asInt }

            val buildResponse: JsonObject = client.request("https://papermc.io/api/v2/projects/paper/versions/1.17.1/builds/$latest")
            val latestChecksum =
                buildResponse
                    .getAsJsonObject("downloads")
                    .getAsJsonObject("application")
                    .get("sha256")
                    .asString

            if (sha256(paperjar) != latestChecksum) {
                println("Checksum doesn't match latest, downloading to verify.")
                download("https://papermc.io/api/v2/projects/paper/versions/1.17.1/builds/$latest/downloads/paper-1.17.1-$latest.jar", paperjar, false)
            }

        } catch (e: Exception) {
            ExceptionDispensary.report(e, "updating")
        }
    }

    private fun getSpaceConnection(): URLConnection {
        val basicAuthenticationEncoded =
            Base64.getEncoder().encodeToString("$spaceUser:$spacePassword".toByteArray(charset("UTF-8")))
        val url =
            URL("https://maven.pkg.jetbrains.space/reflectednetwork/p/internalapi/maven/com/reflectednetwork/RfnetAPI/$nextVer/RfnetAPI-$nextVer.jar")
        val urlConnection = url.openConnection()
        urlConnection.setRequestProperty("Authorization", "Basic $basicAuthenticationEncoded")
        return urlConnection
    }
}